core: add bulk session revocation

cdmx-in opened the pull request:
# Pull Request: Add Bulk Session Revocation Functionality for Admins

## Details

This PR adds bulk session revocation functionality for Admins.
`UserBulkRevokeSessionsForm` added that allows administrators to revoke all active sessions for selected users, with a confirmation dialog to prevent accidental revocation.

### Changes Made

- Created new `UserBulkRevokeSessionsForm.ts` component for handling  this bulk session revocation
- Updated `UserListPage.ts` to integrate the new bulk action with the user list
- Added user-friendly confirmation workflow
<img width="1493" height="775" alt="image" src="https://github.com/user-attachments/assets/68a63681-d7f5-4d26-b35b-4ec959124b0c" />
<img width="715" height="138" alt="image" src="https://github.com/user-attachments/assets/0135c579-4144-477d-bc68-83ca96cce170" />

**Files Changed:**
- `web/src/admin/users/UserBulkRevokeSessionsForm.ts` (+207 lines)
- `web/src/admin/users/UserListPage.ts` (+82/-38 lines)

---

cdmx-in committed:
feat: add bulk session revocation functionality for users

rissson commented on web/src/admin/users/UserBulkRevokeSessionsForm.ts:
```diff
@@ -0,0 +1,207 @@
+                        this.revokedCount++;
+                    }
+                }
+            }
```
Yeah. The UI should call a single endpoint in the backend with the list of users for which to revoke sessions.

> having this directly in the UI makes it much faster

I doubt it. If you have to revoke 200 sessions, you're making 200 API calls (which the associated overhead), instead of 1.

cdmx-in replied:
Have made the necessary changes, @rissson can you please review onces

cdmx-in committed:
Merge branch 'goauthentik:main' into revoke-user-sessions-ui

cdmx-in committed:
feat: add bulk delete functionality for authenticated sessions

- Implemented BulkDeleteSessionSerializer for handling bulk session deletions.
- Added bulk_delete action to AuthenticatedSessionViewSet for revoking sessions by user IDs.
- Updated API schema to include new endpoint for bulk session deletion.
- Modified UserBulkRevokeSessionsForm to utilize the new bulk delete API.

rissson commented on authentik/core/api/authenticated_sessions.py:
```diff
@@ -115,3 +123,14 @@ class AuthenticatedSessionViewSet(
     ordering = ["user__username"]
     owner_field = "user"
+
+    @action(detail=False, methods=["POST"], pagination_class=None, filter_backends=[])
```
We need to check for permissions here. `authentik_core.delete_authenticatesession` IIRC

rissson commented on authentik/core/api/authenticated_sessions.py:
```diff
@@ -115,3 +123,14 @@ class AuthenticatedSessionViewSet(
+    def bulk_delete(self, request: Request) -> Response:
+        """Bulk revoke all sessions for multiple users"""
+        serializer = BulkDeleteSessionSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
```
```suggestion
    @validate(BulkDeleteSessionSerializer)
    def bulk_delete(self, request: Request, body: BulkDeleteSessionSerializer) -> Response:
        """Bulk revoke all sessions for multiple users"""
```
You can then use `body.validated_data`.
The import is `from authentik.api.validation import validate`

rissson replied:
Also, if we change this to DELETE, the following would be correct:
```suggestion
    @validate(BulkDeleteSessionSerializer, location="query")
    def bulk_delete(self, request: Request, query: BulkDeleteSessionSerializer) -> Response:
        """Bulk revoke all sessions for multiple users"""
```

rissson commented on authentik/core/api/authenticated_sessions.py:
```diff
@@ -115,3 +123,14 @@ class AuthenticatedSessionViewSet(
     owner_field = "user"
+
+    @action(detail=False, methods=["POST"], pagination_class=None, filter_backends=[])
+    def bulk_delete(self, request: Request) -> Response:
```
This should have an `@extend_schema` decorator for correct request and response schema

rissson commented on authentik/core/api/authenticated_sessions.py:
```diff
@@ -115,3 +123,14 @@ class AuthenticatedSessionViewSet(
+        serializer.is_valid(raise_exception=True)
+
+        user_ids = serializer.validated_data.get("user_ids", [])
+        deleted_count, _ = AuthenticatedSession.objects.filter(user_id__in=user_ids).delete()
```
```suggestion
        user_pks = body.validated_data.get("user_pks", [])
        deleted_count, _ = AuthenticatedSession.objects.filter(user_pk__in=user_pks).delete()
```
With the other changes

cdmx-in committed:
Update authentik/core/api/authenticated_sessions.py

Co-authored-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>
Signed-off-by: CodeMax IT Solutions Pvt. Ltd. <137166088+cdmx-in@users.noreply.github.com>

cdmx-in committed:
Update authentik/core/api/authenticated_sessions.py

PassiveSerializer for BulkDeleteSessionSerializer

Co-authored-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>
Signed-off-by: CodeMax IT Solutions Pvt. Ltd. <137166088+cdmx-in@users.noreply.github.com>

cdmx-in committed:
Update authentik/core/api/authenticated_sessions.py

user_pks  instead of user_ids

Co-authored-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>
Signed-off-by: CodeMax IT Solutions Pvt. Ltd. <137166088+cdmx-in@users.noreply.github.com>

cdmx-in committed:
feat: enhance bulk delete functionality for authenticated sessions

cdmx-in commented:
@rissson, have made the requested changes, please review

cdmx-in committed:
feat: update bulk delete endpoint for authenticated sessions to use DELETE method and query parameters

cdmx-in committed:
Merge branch 'goauthentik:main' into revoke-user-sessions-ui

rissson reviewed (changes requested):
Just a minor nit, otherwise the backend looks good to me!

cdmx-in committed:
Update authentik/core/api/authenticated_sessions.py

Co-authored-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>
Signed-off-by: CodeMax IT Solutions Pvt. Ltd. <137166088+cdmx-in@users.noreply.github.com>

cdmx-in commented:
> Just a minor nit, otherwise the backend looks good to me!

Done. :)

cdmx-in committed:
Merge branch 'goauthentik:main' into revoke-user-sessions-ui

cdmx-in committed:
Merge branch 'goauthentik:main' into revoke-user-sessions-ui

cdmx-in commented:
Requesting a review for this PR, Please review and merge

rissson committed:
Merge branch 'main' into revoke-user-sessions-ui

rissson committed:
lint

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson committed:
Merge branch 'main' into revoke-user-sessions-ui

dewi-tik committed:
Merge branch 'main' into revoke-user-sessions-ui

dewi-tik committed:
Prettier

chesachesa047-droid commented on web/src/admin/users/UserListPage.ts:
```diff
@@ -163,45 +164,50 @@ export class UserListPage extends WithBrandConfig(
+                </button>
+            </ak-forms-delete-bulk>`;
     }
 
```
![image](https://github.com/user-attachments/assets/7929015c-f93e-4e61-be46-7e8555104ff6)
